File: LR1_GO/Stack.go
package main

import "fmt"

func SPUSH(stack *Stack, value string) {
	newNode := &SNode{Data: value, Next: stack.Top}
	stack.Top = newNode
	stack.Size++
}

func SPOP(stack *Stack) {
	if stack.Top == nil {
		fmt.Println("Стек пуст!")
		return
	}
	
	stack.Top = stack.Top.Next
	stack.Size--
}

func SPRINT(stack *Stack) {
	if stack.Top == nil {
		fmt.Println("Стек пуст!")
		return
	}
	
	fmt.Printf("Стек (%d): ", stack.Size)
	curr := stack.Top
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func SCLEAR(stack *Stack) {
	stack.Top = nil
	stack.Size = 0
	fmt.Println("Стек очищен")
}

File: LR1/Stack.cpp
#include "Stack.h"
#include <iostream>
#include <fstream>

using namespace std;

void SPUSH(Stack& stack, const string& value) {
    SNode* newNode = new SNode(value, stack.top);
    stack.top = newNode;
    stack.size++;
}

void SPOP(Stack& stack) {
    if (!stack.top) { 
        cout << "Стек пуст!\n"; 
        return; 
    }
    SNode* temp = stack.top;
    stack.top = stack.top->next;
    delete temp;
    stack.size--;
}

void SPRINT(const Stack& stack) {
    if (!stack.top) { 
        cout << "Стек пуст!\n"; 
        return;
    }
    cout << "Стек (" << stack.size << "): ";
    SNode* curr = stack.top;
    while (curr) {
        cout << curr->data;
        if (curr->next) cout << " -> ";
        curr = curr->next;
    }
    cout << endl;
}

void SCLEAR(Stack& stack) {
    while (stack.top) {
        SNode* temp = stack.top;
        stack.top = stack.top->next;
        delete temp;
    }
    stack.size = 0;
    cout << "Стек очищен" << endl;
}
